"""Code generated by Speakeasy (https://speakeasyapi.dev). DO NOT EDIT."""

from __future__ import annotations
import dataclasses
from ..shared import scanconfigerror as shared_scanconfigerror
from dataclasses_json import Undefined, dataclass_json
from enum import Enum
from sdk import utils
from typing import Optional

class ScanRunErrorTraceCodeEnum(str, Enum):
    r"""Output only. Indicates the error reason code."""
    CODE_UNSPECIFIED = 'CODE_UNSPECIFIED'
    INTERNAL_ERROR = 'INTERNAL_ERROR'
    SCAN_CONFIG_ISSUE = 'SCAN_CONFIG_ISSUE'
    AUTHENTICATION_CONFIG_ISSUE = 'AUTHENTICATION_CONFIG_ISSUE'
    TIMED_OUT_WHILE_SCANNING = 'TIMED_OUT_WHILE_SCANNING'
    TOO_MANY_REDIRECTS = 'TOO_MANY_REDIRECTS'
    TOO_MANY_HTTP_ERRORS = 'TOO_MANY_HTTP_ERRORS'
    STARTING_URLS_CRAWL_HTTP_ERRORS = 'STARTING_URLS_CRAWL_HTTP_ERRORS'


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class ScanRunErrorTrace:
    r"""Output only. Defines an error trace message for a ScanRun."""
    
    code: Optional[ScanRunErrorTraceCodeEnum] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.get_field_name('code'), 'exclude': lambda f: f is None }})
    r"""Output only. Indicates the error reason code."""  
    most_common_http_error_code: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.get_field_name('mostCommonHttpErrorCode'), 'exclude': lambda f: f is None }})
    r"""Output only. If the scan encounters TOO_MANY_HTTP_ERRORS, this field indicates the most common HTTP error code, if such is available. For example, if this code is 404, the scan has encountered too many NOT_FOUND responses."""  
    scan_config_error: Optional[shared_scanconfigerror.ScanConfigError] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.get_field_name('scanConfigError'), 'exclude': lambda f: f is None }})
    r"""Defines a custom error message used by CreateScanConfig and UpdateScanConfig APIs when scan configuration validation fails. It is also reported as part of a ScanRunErrorTrace message if scan validation fails due to a scan configuration error."""  
    